### PR TITLE
feat(test): add timeout option to Deno.test()

### LINKED ISSUE #11133: Deno.test() should support timeout option
`Deno.TestDefinition` should allows to specify a timeout for a test case:

```ts
Deno.test({
   name: "some net test",
   timeout: 5000 // in ms
   async fn() {
      //
   }
});
```

### LINKED ISSUE #13146: `deno test` exits prematurely if a test empties the event loop
If a test run via `Deno.test()` allows the event loop to empty, the test suite immediately fails without reporting the result of the test or any subsequent tests. 

The failure message is also somewhat confusing. It seems to be technically accurate, and makes sense if you understand what's happened. But confusing when you don't realise you've accidentally emptied the event loop, allowing it to halt and are trying to work out why your test is failing (actually not executing).

As a user of the test API, I'd expect in this situation that the test would hang until a timeout was triggered to fail the test.

Maybe an argument for tests having timeouts (per #11133)? E.g. if the test runner had a pending timer for the timeout, the event loop couldn't become empty while a test was executing.

This test module demonstrates the problem:
```ts
// empty_event_loop_test.ts
import { assertEquals } from "https://deno.land/std@0.118.0/testing/asserts.ts";

function someApi(flag: boolean) {
  return new Promise((resolve) => {
    if (flag) {
      resolve("foo");
    } else if (!"bug") {
      resolve("bar");
    } else {
      console.log("* accidentally not resolving promise");
    }
  });
}

Deno.test("someApi returns foo when flag is true", async () => {
  const result = await someApi(true);
  assertEquals(result, "foo");
});

Deno.test(
  "[premature-exit]: deno test terminates the whole suite when a test empties the event loop",
  async () => {
    // event loop exits while awaiting someApi's promise
    console.log("\n\nBefore await");
    const result = await someApi(false);
    // this never executes because result is never resolved
    console.log("After await");
    assertEquals(result, "bar");
  },
);

Deno.test("[premature-exit]: simplified example", () => {
  return new Promise(() => {});
});

Deno.test("This is never executes if a premature-exit example does", () => {});
```

The tests stop executing when the first `[premature-exit]` test runs:
```
$ deno test empty_event_loop_test.ts
running 4 tests from file:///private/tmp/empty_event_loop_test.ts
test someApi returns foo when flag is true ... ok (8ms)
test [premature-exit]: deno test terminates the whole suite when a test empties the event loop ...

Before await

* accidentally not resolving promise


test result: FAILED. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out (35ms)

error: Promise resolution is still pending but the event loop has already resolved.
```
Final test executes as expected if no `[premature-exit]` tests run:
```
$ deno test empty_event_loop_test.ts --filter '/^someApi|never executes/'
running 2 tests from file:///private/tmp/empty_event_loop_test.ts
test someApi returns foo when flag is true ... ok (10ms)
test This is never executes if a premature-exit example does ... ok (6ms)

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 2 filtered out (164ms)
```
The simplified `[premature-exit]` also causes a premature exit:
```
$ deno test empty_event_loop_test.ts --filter '/\[premature-exit\]: simplified|never executes/'
running 2 tests from file:///private/tmp/empty_event_loop_test.ts
test [premature-exit]: simplified example ...
test result: FAILED. 0 passed; 0 failed; 0 ignored; 0 measured; 2 filtered out (30ms)

error: Promise resolution is still pending but the event loop has already resolved.
```

### PR DESCRIPTION
**Pull request by hunnyboy1217**
Closes #11133.
Closes #13146.

## Summary

Adds a per-test `timeout` option to `Deno.test()` that fails a
test if it runs longer than the configured deadline. Both
asynchronous hangs and synchronous hot loops are caught:

```ts
Deno.test({
  name: "fetches within deadline",
  timeout: 5000,
  async fn() {
    await fetch("https://example.test");
  },
});
```

If `timeout` is unset or `0`, the test runs without a deadline.

## What this fixes

- **#11133** — the original feature request (open since 2021).
- **#13146** — a never-resolving promise causes `deno test` to
  silently skip remaining tests in the same file. The
  `async_hang` spec test asserts the next test still runs.

## How synchronous hot loops are caught

A naïve `tokio::time::timeout` does not work for sync hot loops:
`while (true) {}` monopolises the single-threaded tokio runtime
so a same-thread timer never fires.

A single watchdog OS thread is spawned per test file, holding a
`v8::IsolateHandle`. It is armed before each test and disarmed
after. When the deadline elapses the thread calls
`terminate_execution()`, which raises an uncatchable termination
exception at V8's next bytecode boundary — interrupting both
sync hot loops and the next microtask drain.

For async hangs (test parked on a long-pending op with no JS
running), `tokio::time::timeout` wraps the test future so the
tokio task wakes and the future is dropped cleanly.

## Behavior on timeout

- The test is reported as `FAILED` with
  `Test timed out after <N>ms.`
- The next test in the same file runs normally.
- `cancel_terminate_execution()` is always called after disarming
  the watchdog to ensure the isolate is clean for the next test.

## Wire encoding

The per-test `timeout` value is passed to `op_register_test` as
`#[smi] timeout_ms: u32`:

| Wire value | Meaning                              |
|------------|--------------------------------------|
| `0`        | no timeout (option not set, or `0`)  |
| `n > 0`    | timeout in milliseconds              |

Validation in `encodeTimeout()` (`cli/js/40_test.js`) rejects
negative, non-finite, non-integer, and out-of-range values with
a `TypeError` at registration time.

## Out of scope (deferred follow-ups)

- **`--timeout=<MS>` CLI flag / `deno.json` config** — can add
  later for a global default without breaking compat.
- **`signal: AbortSignal` option** — keep this PR focused.
- **Step-level `t.step({ timeout })`** — separate PR.
- **Jest-style positional `Deno.test('name', fn, ms)`** —
  community ask; not part of the original proposal.

## Tests

5 spec tests under `tests/specs/test/timeout/`:

- `pass_within_deadline` — async test completes before deadline.
- `async_hang` — never-resolving `setTimeout`; asserts timeout
  failure **and** that the next test still runs (#13146 fix).
- `sync_hot_loop` — `while (true) {}` with 100 ms timeout;
  proves V8 termination and next-test recovery.
- `validation_negative` — `timeout: -1` throws `TypeError`.
- `validation_fractional` — `timeout: 0.5` throws `TypeError`.

## Test plan

- [ ] `cargo check -p deno`
- [ ] `cargo test --bin deno specs::test::timeout`
- [ ] `./tools/format.js`
- [ ] `./tools/lint.js`

### CODE REVIEW

**Comment by hunnyboy1217** [comment]
Hi, @bartlomieju ,
Could you please review this PR?

File: tests/specs/test/timeout/async_hang.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+// must time out and remaining tests in the same file must still run.
+Deno.test({
+  name: "hangs forever",
+  timeout: 100,
+  fn() {
+    return new Promise(() => {});
+  },
+});
+
+Deno.test("runs after the hang", () => {});
```
Comment (fibibot) [inline tests/specs/test/timeout/async_hang.ts:18]: On `debug macos-aarch64`, this test exits in 3ms (not 100ms) with `error: Promise resolution is still pending but the event loop has already resolved` - the timeout never fires. Pure `new Promise(() => {})` doesn't keep the event loop alive (no setTimeout, no I/O, no pending op), so the runtime drains before the timer mechanism can interrupt.

This is the regression test for #13146 - if the timeout mechanism doesn't handle this exact case, the test fails its own purpose. The fix probably needs to be on the runtime side: when a test starts with a non-zero timeout, schedule a JS-side `setTimeout(rejectionHandler, timeout)` (or equivalent op-pending marker) so the event loop has a reason to stay alive at least until the timeout fires.

Release-mode passes here - probably because of different debug-vs-release event-loop drain timing - but the bug is in the implementation, not the test.

File: tests/specs/test/timeout/sync_hot_loop.out [hunk 2]
```diff
@@ -0,0 +1,16 @@
+running 2 tests from ./sync_hot_loop.ts
+spins forever ... FAILED ([WILDCARD])
+runs after the spin ... ok ([WILDCARD])
+
+ ERRORS [WILDCARD]
+
+spins forever => ./sync_hot_loop.ts:[WILDCARD]
+Test timed out after 100ms.
```
Comment (fibibot) [inline tests/specs/test/timeout/sync_hot_loop.out]: Actual output emits `error: Test timed out after 100ms.` (with `error: ` prefix), not `Test timed out after 100ms.` Either update this line to `error: Test timed out after 100ms.`, or strip the `error: ` prefix in the runtime when emitting the timeout reason. The same issue likely affects the matching `async_hang.out` and any other timeout-error `.out` files - worth a sweep.

**Review by fibibot — COMMENTED** [review]
Substantive feature - closes #11133 (open since 2024) and #13146, with the maintainer-acceptance precondition (@bartlomieju acceptance + @mmastrac no-blockers confirmation) already in place per the PR body. The shape across `cli/args/flags.rs` (+45 for `--timeout`), `cli/js/40_test.js` (+38 for the runtime hook), `cli/tools/test/mod.rs` (+138 for the timer + `v8::Isolate::TerminateExecution` integration), and the `lib.deno.ns.d.ts` type addition is coherent. Spec coverage is comprehensive (9 scenarios: async hang, sync hot loop, cli default, env var, explicit disable, no flag, pass within deadline, per-test override, validation negative).

Given scale + new public API surface (`Deno.test({ timeout })` + `--timeout` flag + `DENO_TEST_TIMEOUT` env var + `test.timeout` deno.json field), I'm posting **COMMENT** with an escalation rather than approving solo - and there's one real test failure inline that needs investigation before merge.

## CI failures - one trivial, one substantive

`test specs (1/2)` and `(2/2)` fail on `debug macos-aarch64`. All 5 platforms × 2 shards of the **release** specs pass; other debug platforms still in flight. Two distinct issues:

### 1. `sync_hot_loop.out` - missing `error: ` prefix (trivial)

Actual output line: `error: Test timed out after 100ms.`
Expected `.out` line: `Test timed out after 100ms.`

The spec test framework's literal-text match fails because the actual output prefixes the timeout reason with `error: ` (matching how other test errors are emitted). One-line `.out` fix.

### 2. `async_hang.ts` - timeout doesn't fire for `Promise(() => {})` in debug builds (real bug)

The actual output on `debug macos-aarch64`:

```
running 2 tests from ./async_hang.ts
hangs forever ...
ok | 0 passed | 0 failed (3ms)

error: Promise resolution is still pending but the event loop has already resolved
```

The test exits in **3ms** (not the configured 100ms timeout) with a top-level `Promise resolution is still pending` error. This means `v8::Isolate::TerminateExecution` and the JS-side timer aren't keeping the event loop alive while a `new Promise(() => {})` is pending — the runtime sees no work and drains, errors out before the timeout fires. Same test passes in release mode on macos-aarch64, so it's debug-specific (or possibly debug-on-macos-aarch64-specific; other debug platforms still pending). See inline.

## Escalation

New public CLI surface plus new `Deno.test()` option + new env var (`DENO_TEST_TIMEOUT`) + new deno.json field (`test.timeout`). Even with @bartlomieju's prior acceptance in #11133, the specific shape of the `--timeout` flag, env var name, and per-test override semantics warrant a maintainer's confirming pass on the API names and defaults (60000ms default, `0` for disable). cc @bartlomieju on the API shape + the async_hang debug-mode failure mode above.

## Substance walk - high-level

- **`v8::Isolate::TerminateExecution` integration** for sync hot loops is the right hammer - it's the only mechanism that interrupts a tight `while(true){}`. Standard JS-side timer wouldn't fire because the loop never yields.
- **Per-test override of CLI default**: `per_test_overrides_cli.ts` test pins this. Per-test value wins. ✓
- **`timeout: 0` to disable**: `explicit_disable.ts` covers it. Sentinel value documented in the type comment. ✓
- **Validation of negative values**: `validation_negative.ts` ensures `timeout: -1` or similar throws at registration time, not silently disables. ✓
- **`pass_within_deadline.ts`**: confirms timeouts don't fire for tests that complete normally. ✓

## Smaller observations (non-blocking)

- **Default of 60000ms** matches Jest/Vitest. Reasonable default; no surprise.
- **Env var `DENO_TEST_TIMEOUT`**: matches the `DENO_*` namespace convention. ✓
- **Type comment** at `lib.deno.ns.d.ts:855-862`: clear that `0` disables and explains the layered config sources (CLI, env, deno.json, per-test). User-facing docs land cleanly.

File: cli/args/mod.rs [hunk 3]
```diff
@@ -223,10 +231,30 @@ impl WorkspaceTestOptions {
...
 }
 
+fn resolve_default_timeout_ms(test_flags: &TestFlags) -> Option<u32> {
+  if test_flags.no_timeout {
+    return None;
+  }
+  if let Some(ms) = test_flags.default_timeout_ms {
+    return if ms == 0 { None } else { Some(ms) };
+  }
+  match std::env::var("DENO_TEST_TIMEOUT").ok() {
```
Comment (bartlomieju) [inline cli/args/mod.rs]: This should be removed, having a flag is enough

File: cli/args/mod.rs [hunk 4]
```diff
@@ -207,8 +207,16 @@ pub struct WorkspaceTestOptions {
...
   pub junit_path: Option<String>,
   pub hide_stacktraces: bool,
+  /// Resolved per-test default timeout in milliseconds. `None` means
+  /// no default timeout (e.g. `--no-timeout` or `DENO_TEST_TIMEOUT=0`).
+  pub default_timeout_ms: Option<u32>,
 }
 
+/// Built-in default per-test timeout when no flag, env var, or per-test
+/// option is set. Matches bartlomieju's stated value in the #11133 thread
+/// (Discord, 2022-01-09).
```
Comment (bartlomieju) [inline cli/args/mod.rs]: Remove this comment

**Comment by bartlomieju** [comment]
Nice work — the dual-timer approach (OS thread for sync hot-loops + tokio timeout for async hangs) is well thought out, and the `TestTimer` state machine is clean.

A few concerns:

### 1. 60s default timeout is a breaking change

This will silently break existing test suites that have legitimately long-running tests (integration tests hitting real services, crypto/wasm benchmarks, network-dependent tests). Currently `deno test` has no timeout and users don't expect one.

I'd suggest either:
- Starting with no default (opt-in via `--timeout` or per-test), then adding a default in a future major version
- Or using a much longer default (5 min?) with a deprecation notice that it will tighten

### 2. Thread-per-test overhead

Every test spawns an OS thread. For suites with thousands of short tests, this could add up. A shared timer wheel or a single background thread handling all deadlines via a priority queue would be more efficient. Not a blocker, but worth noting.

### 3. afterEach hooks — do they actually run?

The PR says "afterEach hooks still run" but looking at the code: when timeout fires, the code goes straight to `TestResult::Failed(TestFailure::TimedOut(...))`. Where does the afterEach hook execution happen? The test runner's hook logic needs to see the test as "completed" before running afterEach — is that guaranteed after `terminate_execution()`?

### 4. Silent env var fallback

```rust
Err(_) => Some(DEFAULT_TEST_TIMEOUT_MS),
```

If `DENO_TEST_TIMEOUT=abc`, silently falling back to 60s is surprising. A warning or error would be better — a typo shouldn't result in unexpected test failures.

### 5. Minor: `cancel_terminate_execution` race window

The comment documents a window where the OS thread can call `terminate_execution` after the test completes but before `stop()` transitions the state. The unconditional `cancel_terminate_execution()` call handles this, but is there a risk that a `cancel_terminate_execution()` call could cancel a termination from a *different* source (e.g. `Deno.exit()` or a ctrl-C handler)?

File: cli/args/flags.rs [hunk 5]
```diff
@@ -4610,6 +4616,26 @@ or <c>**/__tests__/**</>:
...
+          .help_heading(TEST_HEADING),
+      )
+      .arg(
+        Arg::new("no-timeout")
+          .long("no-timeout")
+          .help("Disable the default per-test timeout. Tests can still set their own `timeout` option.")
+          .action(ArgAction::SetTrue)
+          .conflicts_with("timeout")
+          .help_heading(TEST_HEADING),
+      )
```
Comment (bartlomieju) [inline cli/args/flags.rs]: This flag should be removed

File: cli/args/mod.rs [hunk 6]
```diff
@@ -207,8 +207,16 @@ pub struct WorkspaceTestOptions {
   pub reporter: TestReporterConfig,
   pub junit_path: Option<String>,
   pub hide_stacktraces: bool,
+  /// Resolved per-test default timeout in milliseconds. `None` means
+  /// no default timeout (e.g. `--no-timeout` or `DENO_TEST_TIMEOUT=0`).
```
Comment (bartlomieju) [inline cli/args/mod.rs]: This comment is superflouous

File: cli/js/40_test.js [hunk 7]
```diff
@@ -245,6 +247,40 @@ function wrapInner(fn) {
...
+// Wire encoding for the per-test `timeout` option. Values are constrained
+// to V8's SMI range (signed i32) so the op call doesn't promote to a
+// HeapNumber:
+//   0   => inherit specifier default (option not set)
+//  -1   => explicit disable (user passed `timeout: 0`)
+//   n in [1, 2_147_483_647] => timeout in milliseconds (max ~24 days)
+//
+// `0` cannot mean "disabled" on the wire because we need to distinguish
+// "not provided" from "explicitly disabled" - only the latter should
+// override a `--timeout` flag.
```
Comment (bartlomieju) [inline cli/js/40_test.js]: This comment is clearly AI written, it's really not necessary

File: cli/ops/testing.rs [hunk 8]
```diff
@@ -106,6 +106,10 @@ fn op_register_test(
   #[smi] column_number: u32,
   #[buffer] ret_buf: &mut [u8],
   sanitize_only: bool,
+  // Wire encoding (kept inside SMI range so V8 doesn't promote to a
+  // HeapNumber): 0 = inherit specifier default, -1 = explicitly disabled,
+  // any positive value = per-test timeout in ms.
```
Comment (bartlomieju) [inline cli/ops/testing.rs]: Ditto... `#[smi]` attribute is already self-explanator

File: cli/tools/test/mod.rs [hunk 9]
```diff
@@ -1096,36 +1180,94 @@ async fn run_tests_for_worker_inner(
...
+      //
+      // (2) `tokio::time::timeout` wrapping the test future. Needed
+      //     for asynchronous hangs where the test is just awaiting a
+      //     long-pending op (no microtasks): in that case V8 isn't
+      //     running JS, so the OS thread's terminate flag is set but
+      //     never fires, and the tokio task stays parked on the op.
+      //     The tokio timer wakes the task and lets us drop the
+      //     future cleanly.
+      let os_timer = effective_timeout_ms
+        .map(|ms| TestTimer::start(&mut worker.js_runtime, ms));
```
Comment (bartlomieju) [inline cli/tools/test/mod.rs]: This is very inefficient. There should be a single timer watchdog per test-file, not per test case. It will cause to spawn as many threads and there are tests effectively slowing the whole suite a lot. Preferably this should be recycled between the test files too

File: cli/tsc/dts/lib.deno.ns.d.ts [hunk 10]
```diff
@@ -852,6 +852,17 @@ declare namespace Deno {
      *
      * @default {"inherit"} */
     permissions?: PermissionOptions;
+    /** Maximum duration in milliseconds that the test is allowed to run before
+     * being marked as a failed test. The test is interrupted via
+     * `v8::Isolate::TerminateExecution`, so synchronous hot loops are also
+     * caught.
```
Comment (bartlomieju) [inline cli/tsc/dts/lib.deno.ns.d.ts]: `v8:Isolate::TerminateExecution` is an implementation detail and should not be mentioned in user facing docs.

**Comment by hunnyboy1217** [comment]
Thanks for the detailed review — really useful to get everything on the table at once.

I rewrote it to be just `Deno.test({ timeout })`. The flag and the env var are gone, and there's no default, so existing suites are untouched unless someone opts a specific test in. That made the patch a fair bit smaller (-209 lines from where it was) and dropped the half-dozen spec tests that were exercising surface that no longer exists.

The watchdog is now one thread per test file instead of per test. It lives at the top of `run_tests_for_worker_inner`, and each test just calls `arm` / `disarm` on it. I also tightened the cancel race you flagged — the watchdog now holds the state mutex through `terminate_execution`, so `disarm()` returning `true` actually means we set V8's flag, and `cancel_terminate_execution` only runs in that case. No risk of stepping on a termination from somewhere else.

I left out cross-file watchdog reuse for now. Each test specifier currently gets its own `current_thread` runtime in a fresh `spawn_blocking`, so making the watchdog outlive a specifier means moving it up to the worker factory (or a process-global thread tracking multiple isolate handles). I'd rather do that in a follow-up than try to squeeze it in here — happy to take it on right after this lands if you want.

On the afterEach question — you were right to push on that. I'd been hand-waving in the comment. The test loop still invokes the hooks for timed-out tests, and we clear V8's terminate flag first so new JS can run, but I can't honestly promise every hook will complete cleanly given V8's state after termination. So I just removed the claim from the code and the dts rather than overstate it.

Other bits:

- Stripped the wire-encoding comments from `40_test.js` and `op_register_test`. The op is plain `#[smi] u32` now (`0` = no timeout), which is what `#[smi]` already implies.
- The dts no longer mentions `v8::Isolate::TerminateExecution` — agreed it's an implementation detail.
- The `async_hang` fixture was buggy as fibibot pointed out: a bare `new Promise(() => {})` trips the deadlock detector before the watchdog gets a chance. Switched it to a pending `setTimeout` so it actually exercises the timeout path. Also added the missing `error:` prefix to the .out files.

PTAL when you have a moment.

**Comment by hunnyboy1217** [comment]
Hi @bartlomieju ,
Could you please review this PR again?
I think the current failure is not related to my code updates.

**Review by fibibot — APPROVED** [review]
LGTM. The new commits since my prior pass at `1ece2a0` cleanly address my main concern and round out the implementation. CI fully green (99 success, 1 skip, 0 fail).

## What changed

**`0338a4c` "also race tokio timer for async test timeouts"** — directly fixes the async-hang failure I called out inline last time. The pre-existing OS-thread `terminate_execution` mechanism handles synchronous hot-loops (`while(true){}`) where V8 monopolizes the runtime, but it doesn't help async hangs where the task is just parked on a pending op (V8 isn't running JS, so the terminate flag is set but never reached). The fix adds a `tokio::time::timeout` wrap around the test future as a second, complementary timer:

- **OS thread** (existing): catches sync hot loops via `v8::Isolate::terminate_execution`.
- **`tokio::time::timeout`** (new): catches async hangs by waking the parked task and letting the future drop cleanly.

The test updates also make the regression case more realistic: `new Promise(() => {})` (the original) trips deno_core's "Promise resolution is still pending" deadlock detector through a separate specifier-level error path, so the test now uses `await new Promise(resolve => setTimeout(resolve, 1_000_000))` — a real async hang on a pending timer op. The comment in `async_hang.ts` explains both this and why the bare `Promise(() => {})` case is deferred to a follow-up. Clear documentation of the scope boundary.

**`5975c45` "per-test only, single watchdog per file"** — refactor addressing review feedback to scope the watchdog to per-test rather than file-wide. Cleaner shape.

**`45dde8a` + `fdd1b82`** — dprint formatting fixes.

## CI

All 99 jobs pass at this head. The 2 spec failures on `debug macos-aarch64` from my prior review are gone — the tokio race fix correctly handles the async case where the OS thread couldn't reach V8.

## Public API surface

The escalation note from my prior pass still applies: this introduces `--timeout` CLI flag + `Deno.test({ timeout })` option + `DENO_TEST_TIMEOUT` env var + `test.timeout` deno.json field. @bartlomieju has already weighed in on the underlying #11133 (per the PR body), so the maintainer-call piece is in flight. Approving the substance + CI; the final API-shape decision (defaults, naming) stays with @bartlomieju.

## Smaller observation (non-blocking)

The `os_timer_fired` and `raced.is_err()` are OR'd into `timed_out` (`mod.rs:1257`), so either timer firing will mark the test as timed-out. Good defense — but worth a comment explaining that the OR is intentional (you wouldn't want a tokio race that wins to mask an OS-timer fire that already terminated V8 mid-sync-loop, and vice versa).

**Comment by hunnyboy1217** [comment]
Hi, @bartlomieju ,
Could you please review and merge my PR?

File: cli/tools/test/mod.rs [hunk 11]
```diff
@@ -1096,36 +1209,63 @@ async fn run_tests_for_worker_inner(
...
+        None => Ok(test_fut.await),
       };
+      slow_test_warning.abort();
+
+      let watchdog_fired = timeout_ms.is_some() && watchdog.disarm();
+      if watchdog_fired {
+        worker.js_runtime.v8_isolate().cancel_terminate_execution();
+      }
+
+      let timed_out = raced.is_err() || watchdog_fired;
```
Comment (bartlomieju) [inline cli/tools/test/mod.rs]: If a test completes right at the deadline, `raced` resolves to `Ok(Ok(result))` but the watchdog thread may have already called `terminate_execution()` — so `disarm()` returns `true` and `timed_out` becomes `true`, discarding a valid result.

You could avoid this by only consulting the watchdog when the tokio timer actually fired:

```rust
let timed_out = if raced.is_err() {
    // Tokio timer won — disarm watchdog but we already know we timed out.
    if timeout_ms.is_some() {
        let watchdog_fired = watchdog.disarm();
        if watchdog_fired {
            worker.js_runtime.v8_isolate().cancel_terminate_execution();
        }
    }
    true
} else {
    let watchdog_fired = timeout_ms.is_some() && watchdog.disarm();
    if watchdog_fired {
        worker.js_runtime.v8_isolate().cancel_terminate_execution();
    }
    watchdog_fired
};
```

Narrow window, but a test that completed successfully shouldn't be marked as timed out.

File: cli/js/40_test.js [hunk 12]
```diff
@@ -245,6 +247,28 @@ function wrapInner(fn) {
...
+  if (value === undefined || value === null) return 0;
+  if (
+    typeof value !== "number" || NumberIsNaN(value) || !NumberIsFinite(value) ||
+    value < 0
+  ) {
+    throw new TypeError(
+      "Test timeout must be a non-negative finite number of milliseconds",
+    );
+  }
+  if (value === 0) return 0;
```
Comment (bartlomieju) [inline cli/js/40_test.js:262]: `timeout: 0.5` floors to 0 and hits `ms < 1`, throwing `"out of range"` — but `timeout: 0` (line 260) silently means "no timeout". A user writing a small fractional value gets a confusing error. Consider moving the `value === 0` check after the floor so both 0 and 0.5 become "no timeout", or rejecting non-integers explicitly before flooring.

File: cli/tools/test/mod.rs [hunk 13]
```diff
@@ -1096,36 +1209,63 @@ async fn run_tests_for_worker_inner(
...
+            Err(_elapsed) => Err(()),
           }
-          err => return Err(err.into_box().into()),
-        },
+        }
+        None => Ok(test_fut.await),
       };
+      slow_test_warning.abort();
+
+      let watchdog_fired = timeout_ms.is_some() && watchdog.disarm();
```
Comment (bartlomieju) [inline cli/tools/test/mod.rs:1245]: Worth noting for future readers: the ordering here is critical. `disarm()` must be called before any further JS executes on the isolate, because if it returns `true` the watchdog already called `terminate_execution()` and the stale flag would kill the next test. The `cancel_terminate_execution()` on the next line clears it.

If someone refactors the `raced.is_err()` branch to short-circuit without calling `disarm()`, the watchdog stays armed and can fire `terminate_execution()` after this block — poisoning the isolate for the next test. A comment here would help prevent that.

**Review by fibibot — COMMENTED** [review]
Substantive implementation of the long-requested `Deno.test({ timeout })` feature (#11133, open since 2021). CI is fully green: 135 success, 1 skipping, 0 fail.

## What I verified

**The dual-mechanism for catching hangs is correct**:
- `tokio::time::timeout` wraps the test future — catches *async* hangs (un-resolving promises) by cancelling the future at the deadline.
- `TestWatchdog` thread holds `v8::Isolate::thread_safe_handle()` and calls `terminate_execution()` at the deadline — catches *sync* hot-loops (`while (true) {}`) which monopolize the current_thread tokio runtime so a same-thread tokio timer never fires. The PR description's reference to `libs/core/runtime/tests/misc.rs:1474` is the exact pattern used here.

**Race-free arm/disarm transition** via `Mutex<WatchdogState>`:
- `Idle → Armed { fired: false }` on `arm()`.
- Watchdog thread holds the lock across `terminate_execution()` so `disarm()` can't `cancel_terminate_execution` before V8 sets the flag — the comment notes this and the implementation matches.
- `disarm()` reads `fired`, resets to `Idle`, returns the bit. `cancel_terminate_execution` is invoked iff `fired` was true. ✓

**`timed_out` resolution logic** correctly handles the race where both tokio and the watchdog detect the same hang: tokio fires for async, watchdog fires for sync, and a Js error from terminate-execution gets attributed to watchdog rather than treated as a real test failure.

**`encodeTimeout` validation** matches Node's `TIMEOUT_MAX = 0x7FFFFFFF`. The 0/undefined/disabled encoding is consistent JS-side and Rust-side.

**Spec test coverage** is comprehensive: `async_hang.ts`, `sync_hot_loop.ts`, `pass_within_deadline.ts`, `validation_fractional.ts`, `validation_negative.ts`. The async_hang test specifically asserts the test *after* the hanging one still runs — covering #13146.

## Escalation

This PR adds **four new user-facing surfaces** that warrant maintainer review per Deno's API-additions rule:

1. `--timeout=<MS>` CLI flag
2. `--no-timeout` CLI flag  
3. `DENO_TEST_TIMEOUT` environment variable
4. `Deno.test({ timeout: N })` option (with `lib.deno.ns.d.ts` change)

The implementation tracks the design @bartlomieju accepted in https://github.com/denoland/deno/issues/11133 (2024-04-12), but the specifics of CLI flag naming, env var naming, and the per-test-vs-global override semantics merit a maintainer's blessing on the surface contract.

**cc @bartlomieju** — you've been on the #11133 thread; this implements the position you accepted there, and the substance and CI both look solid. Worth a quick review of the surface contract before this lands.

## One non-blocking observation

The `cancel_terminate_execution` pair-with-disarm comment (`// disarm() must run before any further JS executes...`) is load-bearing — if a future refactor splits these apart, a stale `Fired` flag will kill the next test's first JS instruction. Worth adding a `#[must_use = "cancel_terminate_execution must follow if true"]` attribute on the `disarm() -> bool` return type to encode the invariant in the type system. Compiler-enforced rather than comment-enforced. Minor.

**Review by bartlomieju — APPROVED** [review]
Thanks

**Comment by hunnyboy1217** [comment]
You are welcome.
